### PR TITLE
MARKET-1417 Modal component

### PR DESCRIPTION
**Pull request by Saharij**
Screenshot 2022-10-05 at 22 24 14

### CODE REVIEW

File: components/bl-modal-component/src/index.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import { useEffect } from 'react';
+
+export default function ModalComponent({ component, eventHandlers, pods }) {
+  const { modalVisibility } = component;
```
Comment (StasivPavlo) [inline components/bl-modal-component/src/index.js:4]: add `display`, `classList`, `style`

File: components/bl-modal-component/src/index.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+export default function ModalComponent({ component, eventHandlers, pods }) {
+  const { display, classList, style, disabled, modalVisibility } = component;
+  const { onClose } = eventHandlers;
+
+  const modalContentPod = pods['modalContent'];
+
+  useEffect(() => {
+    const handleEscClick = e => {
+      if (e.keyCode === 27) {
```
Comment (v-excelsior) [inline components/bl-modal-component/src/index.js:13]: Magic number,create constant 'ESCAPE_KEY_CODE'

File: components/bl-modal-component/src/index.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        if (onClose) {
+          onClose({ modalVisibility: false });
+        }
+      }
+    };
+
+    document.addEventListener('keydown', handleEscClick);
+
+    return () => document.removeEventListener('keydown', handleEscClick);
+  }, []);
```
Comment (v-excelsior) [inline components/bl-modal-component/src/index.js:23]: how about create a custom hook useCloseOnEscape and move that logic to the bottom of the component? That could make it more readable.

File: components/bl-modal-component/styles/index.less [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    position: absolute;
+    z-index: 2;
+  }
+
+  .backdrop {
+    position: absolute;
+    width: @bl-customComponent-modal-width;
+    height: @bl-customComponent-modal-height;
+    background-color: @bl-customComponent-modal-backdrop-color;
+    z-index: 1;
```
Comment (v-excelsior) [inline components/bl-modal-component/styles/index.less:26]: add variable for z-index, there could be an collision and we must provide an easy way to solve them.

File: components/bl-modal-component/styles/index.less [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  width: @bl-customComponent-modal-width;
+  height: @bl-customComponent-modal-height;
+  z-index: 100;
+
+  .modal-content {
+    position: absolute;
+    z-index: 2;
```
Comment (v-excelsior) [inline components/bl-modal-component/styles/index.less:18]: same as below

File: components/bl-modal-component/src/index.js [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+  const handleClick = () => {
+    if (onClose) {
+      onClose({ modalVisibility: false });
+    }
```
Comment (v-excelsior) [inline components/bl-modal-component/src/index.js:28]: I'm note sure how it works:

If we don't have a handler - it's not closed by backdrop click
If we have - it's closing.

For me that should works in another way:
If handler exists - we provide used a way to close or not using logic
If handler not exists - we close it every time(or add a property to setup that)

Also there must be an action to close it manually - user must have an option to add a button to pod and close it when he need.

**Review by v-excelsior — CHANGES_REQUESTED** [review]

File: components/bl-modal-component/src/index.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    onClose({ modalVisibility: false });
+  };
+
+  if (!display) {
+    return null;
+  }
+
+  return (
+    <>
+      { visibility &&
```
Comment (v-excelsior) [inline components/bl-modal-component/src/index.js:29]: What if we change modalVisibility from handler? Look like it doesn't work.

File: components/bl-modal-component/src/index.js [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+      className={ cn("bl-customComponent-modal", classList, { "bl-customComponent-modal--disabled": disabled }) }>
+      <div className="backdrop" onClick={ handleClick } />
+      <div className="modal-content">
+        { modalContentPod.render() }
+      </div>
+    </div>
+  );
+}
+
+const useCloseOnEscape = (onClose, setVisibility) => {
```
Comment (v-excelsior) [inline components/bl-modal-component/src/index.js:47]: I'm not sure, but what if we add a property to disable "close on escape" for cases when your want to show modal windows for acceptance of some read information? ("click agree to go to the content of the page" or "click agree if you have read agreement")

File: components/bl-modal-component/src/index.js [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+  };
+
+  if (!visibility) {
+    return null;
+  }
+
+  return (
+    <div
+      style={ style }
+      className={ cn("bl-customComponent-modal", classList, { "bl-customComponent-modal--disabled": disabled }) }>
```
Comment (v-excelsior) [inline components/bl-modal-component/src/index.js:38]: Do we really need "disabled" property? I can't imagine cases where we need to show modal with backdrop and disable all content in it.

File: components/bl-modal-component/styles/index.less [hunk 10]
```diff
@@ -0,0 +1,38 @@
+// you can use LESS variables from the current Theme and app extensions of the UI-Container
```
Comment (v-excelsior) [inline components/bl-modal-component/styles/index.less:1]: This line could be deleted from our components.

Do you agree @Valodya?

**Review by v-excelsior — DISMISSED** [review]

File: components/bl-modal-component/styles/index.less (hunk 10 above)
Comment (Valodya) [reply components/bl-modal-component/styles/index.less]: yes, I agree

File: components/bl-modal-component/src/index.js [hunk 11]
```diff
@@ -0,0 +1,62 @@
+import { useState, useEffect } from 'react';
+
+const { cn } = BackendlessUI.CSSUtils;
+const ESCAPE_KEY_CODE = 27;
```
Comment (Valodya) [inline components/bl-modal-component/src/index.js:4]: add an empty line before the const

File: components/bl-modal-component/src/index.js [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+      onClose({ visibility: false });
+    }
+  };
+
+  if (!visibility) {
+    return null;
+  }
+
+  return (
+    <div
```
Comment (Valodya) [inline components/bl-modal-component/src/index.js:38]: in single line please

File: components/bl-modal-component/src/index.js [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+      if (closeOnEscape && e.keyCode === ESCAPE_KEY_CODE) {
+        setVisibility(false);
+        onClose({ visibility: false });
+      }
+    };
+
+    document.addEventListener('keydown', handleEscClick);
+
+    return () => document.removeEventListener('keydown', handleEscClick);
+  }, []);
```
Comment (Valodya) [inline components/bl-modal-component/src/index.js:61]: the `closeOnEscape ` property must be in the deps list since it can be changed in runtime

**Review by Valodya — APPROVED** [review]

**Pull request merged**